quickwit/quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs

```diff
@@ -214,34 +183,53 @@ impl MappingLeaf {
     }
 }
 
-fn json_type_from_json_value(json_value: &JsonValue) -> JsonType {
-    match json_value {
-        JsonValue::Null => JsonType::Null,
-        JsonValue::Bool(_) => JsonType::Bool,
-        JsonValue::Array(_) => JsonType::Array,
-        JsonValue::Number(_) => JsonType::Number,
-        JsonValue::String(_) => JsonType::String,
-        JsonValue::Object(_) => JsonType::Object,
-    }
-}
-
 fn extract_json_val(
-    json_type: JsonType,
-    named_doc: &mut BTreeMap<String, Vec<JsonValue>>,
+    leaf_type: &LeafType,
+    named_doc: &mut BTreeMap<String, Vec<Value>>,
     field_path: &[&str],
     cardinality: Cardinality,
 ) -> Option<JsonValue> {
     let full_path = field_path.join(".");
     let vals = named_doc.remove(&full_path)?;
     let mut vals_with_correct_type_it = vals
         .into_iter()
-        .filter(|json_val| json_type_from_json_value(json_val) == json_type);
+        .filter_map(|value| value_to_json(value, leaf_type));
     match cardinality {
         Cardinality::SingleValue => vals_with_correct_type_it.next(),
         Cardinality::MultiValues => Some(JsonValue::Array(vals_with_correct_type_it.collect())),
     }
 }
 
+fn value_to_json(value: Value, leaf_type: &LeafType) -> Option<JsonValue> {
```

Comment:
I would write this function a bit differently

```rust
fn value_to_json(value: Value, leaf_type: &LeafType) -> Option<JsonValue> {
    // Make sure the requested type and the value type are consistent.
    match (&value, leaf_type) {
        (Value::Str(_), LeafType::Text(_))
        | (Value::I64(_), LeafType::I64(_))
        | (Value::U64(_), LeafType::U64(_))
        | (Value::F64(_), LeafType::F64(_))
        | (Value::Bool(_), LeafType::Bool(_))
        | (Value::IpAddr(_), LeafType::IpAddr(_))
        | (Value::Date(_), LeafType::DateTime(_))
        | (Value::Bytes(_), LeafType::Bytes(_))
        | (Value::JsonObject(_), LeafType::Json(_)) => Some(
            serde_json::to_value(&matched_value).expect("Json serialization should never fail."),
        ),
        (Value::Date(date_time), LeafType::DateTime(date_time_options)) => {
            let json_value = date_time_options
                .format_to_json(*date_time)
                .expect("Invalid datetime is not allowed.");
            Some(json_value)
        }
        _ => return None, // TODO: log a warning.
    }
}
```

Reply:
This is more compact  :+1:, I wanted more detailed steps in code rather than commenting.